Update domain parsing in network classes

spacedmonkey opened the pull request:
Use `parse_url` and `PHP_URL_HOST` to ensure that domain is a 'localhost' and the port is stripped. 

Trac ticket: https://core.trac.wordpress.org/ticket/21077

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Update domain parsing in network classes

The code has been updated to use the PHP_URL_HOST component when parsing the domain for allow_subdomain_install() method and the cookie_domain property in the WP_Network class. This ensures consistent and accurate extraction of the domain, disregarding any potential path or protocol information.

spacedmonkey committed:
Refactor `allow_subdomain_install` code

Adjusted the `allow_subdomain_install` function by reducing the repetitive use of get_option( 'home' ) in the network.php file. This improves readability and the efficiency of the code.

johnbillion committed:
Merge branch 'trunk' into fix/multisite-login

johnbillion reviewed (approved):
Working nicely, thanks.

johnbillion reviewed (changes requested):
Oh maybe not. The unit tests have exploded. Looks like the fix in `_set_cookie_domain()` needs some more work.

johnbillion committed:
Merge branch 'trunk' into fix/multisite-login

spacedmonkey committed:
Update cookie domain assignment in WP_Network class

An enhancement to the cookie domain assignment logic within the WP_Network class has been made. It now uses the parsed domain, if present, otherwise it falls back to the original domain. Additionally, it strips 'www.' from the start of the cookie domain if necessary.

spacedmonkey committed:
Improve cookie domain parsing logic

The bug in the existing logic for parsing cookie domain in the WP_Network class has been fixed. The check now ensures that the $domain is a string before attempting to use it as the cookie domain, providing a more robust solution against edge cases.

johnbillion commented:
https://core.trac.wordpress.org/changeset/58138
